### ISSUE TITLE
fitheight

### ISSUE THREAD

**Issue by rrveex** [dawsers/hyprscroller#103]
Assuming row mode (default), I'd rather _not_ have out-of-viewport vertical windows. For a couple of versions I dragged a patch to automatically "fit all vertical" after each `admit` / `expel` (somehow like niri), but it's getting cumbersome for all the changes.

Are you planning to implement something like this? (the new `cycleheight` seems to be in this direction)

If not, would you consider a PR?

**Comment by dawsers** [comment]
Let me see if I understand what you want..

You want a mode/option/windowrule that when you create a new window in column mode, fits all the windows in that same column to the height of the screen, and when you delete a window, resizes them again.

The way things work right now is:

1. You can set the default height of a window. Some heights are smaller than the height of the screen, but they are not adaptive; they don't change as you keep adding/deleting windows. You can also call `cycleheight`, but there is the same problem.
2. You can call `fitsize` in *column* mode every time you add a window, and that would create the effect I think you are looking for. I believe that is what you implemented with (probably) a script. The problem there is it works using a *windowrule*, so it's available only with `admit/expel`, but not on manual creation/deletion of windows while in *column* mode.

So you would like a mechanism to auto-size windows per column every time you create/remove a window. That mechanism could be an option, a window rule or a mode. Let me know if I am getting it right.

Now for the motivations (I always want to understand the motivation before adding something):

My critique to what I understand you want: not binding, not meaning I am against implementing something similar, just the starting point for a discussion. In my personal case, I wouldn't enjoy having my windows change height automatically every time I open a new one just to make it fit in the screen. As soon as I have opened more than maybe two windows they become "useless". If I wanted something similar, I would use a real automatic tiler. In fact, the reason why I don't use an auto-tiler is because I don't like having windows that are small to the point of being of little use, and that is why *hyprscroller* doesn't limit the user to for everything in the viewport. So sell me your idea :-)

**Comment by rrveex** [comment]
> just want something like what the title says: separate fitsize into fitwidth and fitheight, as dispatchers.

Actually I just care about `fitheight` :D

My workflow regarding this is quite simple:

- need to see stuff -> open a new window (mostly terminal) to display that stuff
- need to see stuff I remember I have already opened -> move focus left/right until needed window comes into viewport; I can't remember on which column I'd have to also focus up/down in order to find my stuff
- need to see stuff *at the same time* (all in viewport) -> find some windows that don't display that much stuff and "merge" them into only 1 column - this is `fitheight`

My patch does this automatically (I never want to have windows above/below the viewport; I think of "above/below" as other workspaces) for `admit` and `expel`; of course I wouldn't expect this automatic behaviour implemented for everybody, the dispatcher could be called on the same hyprland shortcut for the same effect.

```diff
diff --git a/src/column.cpp b/src/column.cpp
index 5b38378..28f843e 100644
--- a/src/column.cpp
+++ b/src/column.cpp
@@ -388,6 +388,7 @@ void Column::update_width(StandardSize cwidth, double maxw) {
 }
 
 void Column::fit_size(FitSize fitsize, const Vector2D &gap_x, double gap) {
+  bool all_same = false;
   const Box &max = row->get_max();
   reorder = Reorder::Auto;
   ListNode<Window *> *from, *to;
@@ -421,6 +422,10 @@ void Column::fit_size(FitSize fitsize, const Vector2D &gap_x, double gap) {
       }
     }
     break;
+  case FitSize::AllVert:
+    all_same = true;
+    // fallthrough
   case FitSize::All:
     from = windows.first();
     to = windows.last();
@@ -441,13 +446,19 @@ void Column::fit_size(FitSize fitsize, const Vector2D &gap_x, double gap) {
   // screen (max.h) among from->to, and readapt the rest
   if (from != nullptr && to != nullptr) {
     double total = 0.0;
+    int num_windows = 0;
     for (auto c = from; c != to->next(); c = c->next()) {
       total += c->data()->get_geom_h();
+      num_windows++;
     }
     for (auto c = from; c != to->next(); c = c->next()) {
       Window *win = c->data();
       win->set_height_free();
-      win->set_geom_h(win->get_geom_h() / total * max.h);
+      if (all_same) {
+        win->set_geom_h(max.h / num_windows);
+      } else {
+        win->set_geom_h(win->get_geom_h() / total * max.h);
+      }
     }
     auto gap0 = from == windows.first() ? 0.0 : gap;
     from->data()->move_to_top(geom.x, max, gap_x, gap0);
diff --git a/src/dispatchers.cpp b/src/dispatchers.cpp
index f608db1..e68e599 100644
--- a/src/dispatchers.cpp
+++ b/src/dispatchers.cpp
@@ -209,6 +209,9 @@ std::optional<FitSize> parse_fit_size(std::string arg) {
     return FitSize::Visible;
   else if (arg == "all")
     return FitSize::All;
+  else if (arg == "allvert")
+    return FitSize::AllVert;
   else if (arg == "toend")
     return FitSize::ToEnd;
   else if (arg == "tobeg" || arg == "tobeginning")
diff --git a/src/enums.h b/src/enums.h
index 5ba30e8..6acdcc5 100644
--- a/src/enums.h
+++ b/src/enums.h
@@ -12,7 +12,8 @@ enum class Direction {
   Middle,
   Invalid
 };
-enum class FitSize { Active, Visible, All, ToEnd, ToBeg };
+enum class FitSize { Active, Visible, All, ToEnd, ToBeg, AllVert };
 enum class Mode { Row, Column };
 
 #endif // SCROLLER_ENUMS_H
diff --git a/src/row.cpp b/src/row.cpp
index 55a30e5..05ea23c 100644
--- a/src/row.cpp
+++ b/src/row.cpp
@@ -619,6 +619,9 @@ void Row::expel_window_right() {
   if (width == StandardSize::Free)
     width = StandardSize::OneHalf;
 #endif
+  // fit size original column (fewer windows now)
+  active->data()->fit_size(FitSize::AllVert, calculate_gap_x(active), gap);
+
   active = columns.emplace_after(active, new Column(w, width, maxw, this));
   // Initialize the position so it is located after its previous column
   // This helps the heuristic in recalculate_row_geometry()
@@ -740,7 +743,9 @@ void Row::fit_size(FitSize fitsize) {
   if (overview) {
     return;
   }
-  if (mode == Mode::Column) {
+  // if (mode == Mode::Column) {
+  if (mode == Mode::Column || fitsize == FitSize::AllVert) {
     active->data()->fit_size(fitsize, calculate_gap_x(active), gap);
     return;
   }
diff --git a/src/scroller.cpp b/src/scroller.cpp
index f082ea8..d925892 100644
--- a/src/scroller.cpp
+++ b/src/scroller.cpp
@@ -920,6 +920,8 @@ void ScrollerLayout::admit_window_left(WORKSPACEID workspace) {
     return;
   }
   s->admit_window_left();
+  s->fit_size(FitSize::AllVert);
 }
 
 void ScrollerLayout::expel_window_right(WORKSPACEID workspace) {

```

**Comment by dawsers** [comment]
OK, that is simple, done in d26cb058fc190da7b9054d4a8701b025f5fea7db

Now you also have `fitwidth` and `fitheight`.

As for your workflow, I think the `jump` dispatcher could help you. It is usually quicker than having to "find" stuff. I was doing something similar to what you do, and that is why I implemented it.

**Comment by rrveex** [comment]
That was quick, thank you!

Yes, `jump` was the reason I upgraded hyprland (to match the new hyprscroller); now I can take a look at `jump` :)

**Comment by rrveex** [comment]
Yep, `jump` is *very* nice; doesn't make `fitheight` obsolete though, still need that for "have plenty of stuff *in the viewport*"; thanks again!

**Comment by dawsers** [comment]
Cool, I am glad you find it useful!

I always ask about workflow because we all use the plugin in many different ways, and sometimes it may help my own workflow to listen to how other people use it, and to understand better what new features may be useful.
